scripts/sk_cleanup_tenants/script.py

```diff
@@ -0,0 +1,133 @@
+parser.add_argument("--safekeeper-host", required=True, type=str)
+args = parser.parse_args()
+
+access_key = os.getenv("CONSOLE_API_TOKEN")
+endpoint: str = "https://console.stage.neon.tech/api"
+
+trash_dir: Path = args.trash_dir
+dry_run: bool = args.dry_run
+logging.info(f"dry_run={dry_run}")
+sk_id: int = args.safekeeper_id
+sk_host: str = args.safekeeper_host
+
+assert trash_dir.is_dir()
+
+###
+
+
+def console_get(rel_url):
+    r = requests.get(
+        f"{endpoint}{rel_url}",
+        headers={
+            "Authorization": f"Bearer {access_key}",
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        },
+    )
+    r.raise_for_status()
+    return r
+
+
+def tenant_is_deleted_in_console(tenant_id):
+    r = console_get(f"/v1/admin/projects?search={tenant_id}&show_deleted=true")
+    r = r.json()
+    results = r["data"]
+    assert len(results) == 1, f"unexpected results len: {results}"
+    r = results[0]
+    assert r["tenant"] == tenant_id, f"tenant id doesn't match: {r}"
+    assert r["safekeepers"] is not None, f"safekeepers is None: {r}"
+    assert any(sk["id"] == sk_id for sk in r["safekeepers"]), f"safekeeper id not found: {r}"
+    assert "deleted" in r, f"{r}"
+    return r["deleted"] is True
+
+
+def call_delete_tenant_api(tenant_id):
+    r = requests.delete(f"http://{sk_host}:7676/v1/tenant/{tenant_id}")
+    r.raise_for_status()
+    return r
+
+
+def cleanup_tenant(tenant_id):
+
+    tenant_dir = Path(f"/storage/safekeeper/data/{tenant_id}")
+
+    if not tenant_dir.exists():
+        logging.info("tenant directory doesn't exist, assuming it has been cleaned already")
+        return
+
+    if not tenant_is_deleted_in_console(tenant_id):
+        logging.info("tenant is not deleted in console, skipping")
+        return
+
+    logging.info("assertions passed")
+
+    if dry_run:
+        return
+
+    logging.info("deleting tenant")
+
+    tenant_dir_in_trash = trash_dir / tenant_dir.relative_to("/")
+    tenant_dir_in_trash.parent.mkdir(parents=True, exist_ok=True)
+
+    assert not tenant_dir_in_trash.exists(), f"{tenant_dir_in_trash}"
+    assert tenant_dir_in_trash.parent.exists(), f"{tenant_dir_in_trash}"
+    # double-check
+    assert tenant_dir.exists(), f"{tenant_dir}"
+    assert tenant_dir.is_dir(), f"{tenant_dir}"
+
+    logging.info(f"copying {tenant_dir} to {tenant_dir_in_trash}")
+    shutil.copytree(src=tenant_dir, dst=tenant_dir_in_trash, symlinks=False, dirs_exist_ok=False)
+
+    logging.info(f"deleting {tenant_dir}")
+    call_delete_tenant_api(tenant_id)
+
+    logging.info("tenant is now deleted, checking that it's gone")
+    assert not tenant_dir.exists(), f"{tenant_dir}"
+
+
+if os.path.exists("script.pid"):
+    logging.info(
+        f"script is already running, with pid={Path('script.pid').read_text()}. Terminate it first."
+    )
+    exit(1)
+
+with open("script.pid", "w", encoding="utf-8") as f:
```

Comment:
https://github.com/mosquito/python-pidfile 

it also checks that the process is running. For instance, the file is present, but the PID in this file refers to the already killed process.

Possible bug:
1. start process
2. send SIGTERM to this process

Expected result:
the new process can be started

Actual result:
"Sciprt is already running..."

Reply:
PID library is nice, but it's better (and easier) not to use any external libraries in scripts running on the storage instances. That's because we don't have pip installed and it's easier not to require additional steps (such as `pip install pidfile`) before running the script.

> Possible bug
I'm not sure if it's good to allow to run script if it was finished unexpectedly, manually checking logs and pidfile in that case sounds ok.

Reply:
I hope that you have an installed pyenv. Then, it makes sense

Reply:
wait, but how can you use `requests`? it's not a built-in package

Reply:
🤔 

Probably it comes with our debian distro? Or we install something that depends on it in our provisioning scripts? Apparently we have a lot of installed packages (but no pip/pyenv):
```
ssm-user@safekeeper-1:/usr/lib/python3/dist-packages$ ls
Jinja2-2.11.3.egg-info                              colorama-0.4.4.egg-info            oauthlib
MarkupSafe-1.1.1.egg-info                           configobj-5.0.6.egg-info           oauthlib-3.1.0.egg-info
PyJWT-1.7.1.egg-info                                configobj.py                       pkg_resources
PySimpleSOAP-1.16.2.egg-info                        cryptography                       pvectorc.cpython-39-x86_64-linux-gnu.so
PyYAML-5.3.1.egg-info                               cryptography-3.3.2.egg-info        pyasn1
__pycache__                                         curl                               pyasn1-0.4.8.egg-info
_cffi_backend.cpython-39-x86_64-linux-gnu.so        dateutil                           pycurl-7.43.0.6.egg-info
_dbus_bindings.cpython-39-x86_64-linux-gnu.so       dbus                               pycurl.cpython-39-x86_64-linux-gnu.so
_dbus_glib_bindings.cpython-39-x86_64-linux-gnu.so  dbus_python-1.2.16.egg-info        pyrsistent
_distutils_hack                                     deb822.py                          pyrsistent-0.15.5.egg-info
_pyrsistent_version.py                              debconf.py                         pysimplesoap
_version.py                                         debian                             python_apt-2.2.1.egg-info
_yaml.cpython-39-x86_64-linux-gnu.so                debian_bundle                      python_dateutil-2.8.1.egg-info
aiohttp                                             debianbts                          python_debian-0.1.39.egg-info
aiohttp-3.7.4.egg-info                              distro_info-1.0.egg-info           python_debianbts-3.1.0.egg-info
apt                                                 distro_info.py                     reportbug
apt_inst-stubs                                      docutils                           reportbug-7.10.3+deb11u1.egg-info
apt_inst.cpython-39-x86_64-linux-gnu.so             docutils-0.16.egg-info             requests
apt_pkg-stubs                                       httplib2                           requests-2.25.1.egg-info
apt_pkg.cpython-39-x86_64-linux-gnu.so              httplib2-0.18.1.egg-info           roman-2.0.0.egg-info
aptsources                                          idna                               roman.py
async_timeout                                       idna-2.10.egg-info                 rsa
async_timeout-3.0.1.egg-info                        importlib_metadata                 rsa-4.0.egg-info
attr                                                importlib_metadata-1.6.0.egg-info  s3transfer
attrs-20.3.0.egg-info                               jinja2                             s3transfer-0.3.4.egg-info
awscli                                              jmespath                           setuptools
awscli-1.19.1.egg-info                              jmespath-0.10.0.egg-info           setuptools-52.0.0.egg-info
blinker                                             jsonpatch-1.25.egg-info            six-1.16.0.egg-info
blinker-1.4.egg-info                                jsonpatch.py                       six.py
boto                                                jsonpointer-2.0.egg-info           typing_extensions-3.7.4.3.egg-info
boto-2.49.0.egg-info                                jsonpointer.py                     typing_extensions.py
botocore                                            jsonschema                         unattended_upgrades-0.1.egg-info
botocore-1.20.0.egg-info                            jsonschema-3.2.0.egg-info          urllib3
certifi                                             jwt                                urllib3-1.26.5.egg-info
certifi-2020.6.20.egg-info                          lsb_release.py                     validate.py
chardet                                             markupsafe                         yaml
chardet-4.0.0.egg-info                              more_itertools                     yarl
cloud_init-20.4.1.egg-info                          more_itertools-4.2.0.egg-info      yarl-1.6.3.egg-info
cloudinit                                           multidict                          zipp-1.0.0.egg-info
colorama                                            multidict-5.1.0.egg-info           zipp.py
```